did-git-spec.md

```diff
@@ -125,29 +125,36 @@ Example:
   }
 }
 ```
+The commit adding your DID document should be signed by the key associated private key of the public key referenced in the DID document.
 
 ## Read
 
 Notes: since we don't know the repo did string when the repo DID document is created, the repo DID document will not contain the "id" did string for the repo. The `git did read` operation will "resolve" the "id" by looking up the SHA1 hash of the commit that added the DID document to the repo and dynamically add the "id" member to what is rendered to the user.
 
 ## Update 
-- repo.did
-    - updating service_endpoints?
-- [personal-hash].did
-    - is any signature/crypto verification required to update a [personal-hash].did, or, for example, can I change your keys and commit it to the repo to effectively lock you out?  Or is change to .did/[personal-hash].did only allowed where [personal-hash].did matches the signer?
+**1. Repository DID**
+
+There aren't currently any keys associated with the respository, so there will ne no need to update the Public Keys in the DID document.  The process to change the maintainers or the canonical endpoint should be defined in the `.did/governance.md` file.
+
+**2. Contributor DID**
+
+To update a Contributor DID document, remove the previous DID document from the repository and commit.  This commit should be signed by the private key of the DID document being removed.  Then create a new DID document with the appropriate information and commit to the repository.  This commit should be signed by the private key of the DID document being added.
```

Comment:
I was thinking that updates would be like patches to the file instead of a delete and create operation. The main reason is that I imagined people storing credentials and proofs in their DID document. Things like a "maintainer" credential that is signed by M of N maintainers that promotes them to being a maintainer. The credential would be a signature over their public key and their role assignment. When the contributor then updates their public key, it would be nice to just rely on provenance to validate the new public key rather than requiring that all maintainers then re-sign the new key to preserve the contributors maintainer status.

Deleting and creating seems like that may make things harder. What's the reason for delete/create if all they want to do is update their public key?

Reply:
If the DID document receives a public key update, then the contributor ID changes right? So the DID document will need to be renamed to the new `<contributor ID>.did`. Git handles file renames explicitly.

I'm just pointing out a wrinkle in the UPDATE method on the contributor DID string. If the file gets renamed, that would make validating old signatures more difficult since the state of the DID keyring will need to be know at the point in Git history that the signature was made so that the lookup can be done. This isn't a problem but an important implementation detail. We were already heading in this direction when we were talking about deleting repo.did files and "regime changes".

So to be clear, consider the following scenario:
```
A  - - - Alice adds her DID doc with ID of 1234.did
|
B  - - - Alice adds a signed commit with key ID 1234
|
C  - - - Alice updates her key and her ID changes to 5678
|
D  - - - Alice adds a signed commit with key ID 5678
```

At point B in the repo history the `.did` keyring folder looks like:
```
.did/
  + - - - 1234.did
```
So validating the signature of commit B will require knowing what the .did folder looks like a that time in history so that the DID string in the signature (`did:git:<repo ID>/1234`) can be used to read the public key used to verify the signature.

Then again at point D in the repo timeline, the `.did` keyring folder will look like:
```
.did/
  + - - - 5678.did
```
And the DID string in that signature will be `did:git:<repo ID>/5678`.

Reply:
The thought behind delete/add was to prevent someone from altering the file maliciously.  How can we verify that the new set of keys belongs to the person that originally committed the Contributor DID?  I was also thinking the porcelain could assist here with a `rotate` command that would handle the multiple changes.

Seems that the delete/add would lend allow us to automate the check and would make things easier in the long run.  I can reliable prove that I have the proper keys to delete a file and add a file but maybe I'm missing something with respect to proving I can change the file.  Maybe we should move this to the mailing list to solicit feedback from those that may have solved this already.  Thoughts?

Reply:
After a little more thought (at the park with my daughter) I think if `git mv .did/<old key file>.did <new key file>.did` is used on the Contributor DID file and then signed with the key being replaced it would work and also provide the ability to see key rotations.  Assuming that we validate the commits based on the keys we have in our current repository, then the change would be valid prior to the merge and then after the merge the new set of keys would apply.  Thoughts?

Reply:
@twshelton I like the `git mv` signed by both old and new keys. The trick is that the signature with the new key has to be signed by the old key for chaining purposes. I also think the governance should have rules that maintainer key rotations must be confirmed with signatures from some M of N of the other maintainers. But then that opens up an attack from a cabal of maintainers blocking the key rotation of another maintainer or potentially preventing key rotation when time is of the essence due to a breach of security.

Regardless, I think `git mv` is the right way to go.